### PR TITLE
script to gather log information for efficient troubleshooting of software environment

### PR DESCRIPTION
**Pull request by nandlalyadav57**
Added script to gather log information for efficient troubleshooting of software environment.
We are going to use the script to instruct customers to collect logs: https://github.com/astronomer/docs/pull/4026/files
Context: https://astronomer.slack.com/archives/C07G6J17AP6/p1723130966001819

### CODE REVIEW

File: support-utils/cre_astro_software_gather_logs.sh [hunk 1]
```diff
@@ -0,0 +1,392 @@
+#!/bin/bash
+timestamp=`date '+%d/%m/%Y %H:%M:%S'`
+# Put the name of the astronomer ASTRONOMER_NAMESPACE below
+#export ASTRONOMER_NAMESPACE=astronomer
+#export DIR="/tmp"
+#export BASEDOMAIN=nandlal51.astro-cre.com
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:6]: This comment data should be replaced with example.com or the comment should be removed.

https://www.rfc-editor.org/rfc/rfc2606

File: support-utils/cre_astro_software_gather_logs.sh [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+#export BASEDOMAIN=nandlal51.astro-cre.com 
+echo "****Example to Run This Script:****
+
+Enter your Astronomer Namespace Name: (This is the namespace where all core Astronomer components like Houston, Commander, Registry, etc., are running)
+astronomer <============ 1
+Enter the directory path where you want to save your exported log files:
+/tmp     <============ 2
+If you have a test cluster with the URL https://app.cre-software-01.astro-cre.com, then your base domain would be cre-software-01.astro-cre.com.
+Enter your base domain:
+cre-software-01.astro-cre.com  <============ 3
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:15]: This also needs to be replaced with example.com https://www.rfc-editor.org/rfc/rfc2606

File: support-utils/cre_astro_software_gather_logs.sh [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+****
+
+You can refer to the above example as a guide when entering values for your environment."
+
+
+echo "Enter your Astronomer Namespace Name:"
+read ASTRONOMER_NAMESPACE
+echo "Enter the path of directory where you want to keep your log files exported:"
+read DI
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:25]: `DI` is too short and ambiguous.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+export DIR=$DI/\astro_logs
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:26]: Do we need to escape this slash here?

File: support-utils/cre_astro_software_gather_logs.sh [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+echo "Checking ENDPOINTS"
+
+for EP in $(kubectl describe svc kube-dns -n kube-system|grep Endpoints|awk '{print $2}'|uniq|sed 's/,/\n/g'|sed 's/:[^[:blank:]]*//'); do
+echo "======================CHECKING Houston ENDPOINT for $EP======================";nslookup houston.$BASEDOMAIN > $DIR/nslookup_houston_$EP.$BASEDOMAIN.log; echo " you have to run nslookup houston.$BASEDOMAIN $EP inside any of the pods lets say inside a nginx pod" >> $DIR/nslookup_houston_$EP.$BASEDOMAIN.log
+echo "PLEASE NOTE ======== you have to run nslookup houston.$BASEDOMAIN $EP inside any of the pods lets say inside a nginx pod to make sure endpoints are running fine"
+done
+
+for i in {1..10} ;do curl -I  https://registry.$BASEDOMAIN; done  > $DIR/curl_check_registry.$BASEDOMAIN.log
+for i in {1..10} ;do curl -I  https://app.$BASEDOMAIN/; done      > $DIR/curl_check_app.$BASEDOMAIN.log
+for i in {1..10} ;do curl -I  https://install.$BASEDOMAIN/; done  > $DIR/curl_check_install.$BASEDOMAIN.log
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:363]: What are we trying to do here? It seems like a loop is not the right approach.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+ls -lhtr $DIR/$ASTRONOMER_NAMESPACE*
+ls -lhtr
+echo "Please attach the zip or .gz file created in $cdir to the Zendesk ticket for reference. If the file exceeds 50 MB, kindly use a cloud storage service like Google Drive or similar."
+#echo "Timing out for 30 sec for zip file to be present before sending"
+#@timeout /t 30
+#####====================================================================================================================================================#####
+#echo "Sharing the logs via mail for troubleshooting"
+#echo "Here are the Platform logs for troubleshooting $Ticket" | mutt -a "$DIR".zip" -a $DIR"_$(date +%F).tar.gz -s "Platform logs for troubleshooting $Ticket" -- $mail
+#####====================================================================================================================================================#####
+#echo "Here are the Platform logs for troubleshooting $Ticket" | mutt -a "$DIR".zip" -s "Platform logs for troubleshooting $Ticket" -- $mail
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:392]: Why are all of these commented code lines here?

File: support-utils/cre_astro_software_gather_logs.sh [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+#for i in {1..10} ;do curl -I  https://alertmanager.$BASEDOMAIN; done
+
+
+#####====================================================================================================================================================#####
+echo "======================creating GZ and zip files======================"
+#####====================================================================================================================================================#####
+cd "$DIR"
+cd ..
+tar -czvf "astro_logs"_$(date +"%Y_%m_%d_%I_%M_%p").tar.gz "$DIR"
+zip -r "astro_logs"_$(date +"%Y_%m_%d_%I_%M_%p").zip "$DIR"
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:379]: These date formats are TZ agnostic which makes them imprecise. Consider using iso8601, rfc3339, or something close to those.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+      echo "Gathering logs of triggerer in $NS Namespace ";kubectl logs deployment/$Release_Name-triggerer -c triggerer > "$NS/AllPodlogs/morelogs/triggerer $NS.log" -n $NS
+      echo "Gathering logs of pgbouncer in $NS Namespace ";kubectl logs deployment/$Release_Name-pgbouncer -c pgbouncer > "$NS/AllPodlogs/morelogs/pgbouncer _$NS.log" -n $NS    
+      echo "Gathering logs of flower  in $NS Namespace ";kubectl logs deployment/$Release_Name-flower > "$NS/AllPodlogs/morelogs/flower _$NS.log" -n $NS
+      echo "Gathering logs of statsd in $NS Namespace ";kubectl logs deployment/$Release_Name-statsd > "$NS/AllPodlogs/morelogs/statsd_$NS.log" -n $NS
+      echo "Gathering logs of redis in $NS Namespace ";kubectl logs sts/$Release_Name-redis > "$NS/AllPodlogs/morelogs/redis_$NS.log" -n $NS 
+      echo "Gathering helm history in $NS Namespace";helm history $Release_Name -n $NS > "$NS/AllPodlogs/morelogs/helm_history_$Release_Name.yaml"
+      echo "Gathering helm values from $NS Namespace";helm get values $Release_Name -o yaml -n $NS  > "$NS/AllPodlogs/morelogs/helm_values_$Release_Name.yaml"
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:346]: This is a lot of duplicate code that should probably be turned into a function that takes the name of the pod, the namespace, and the release name.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+echo "Gathering Astro version status";astro version  >> "$DIR/Enviornment_Info.log"
+echo "=======================docker version output==========================================================================" >> "$ASTRONOMER_NAMESPACE/Deployment_logs/Enviornment_Info.log"
+echo "Gathering docker version status";docker version  >> "$DIR/Enviornment_Info.log"
+echo "=======================helm version output==========================================================================" >> "$ASTRONOMER_NAMESPACE/Deployment_logs/Enviornment_Info.log"
+echo "Gathering helm version status";helm version  >> "$DIR/Enviornment_Info.log"
+echo "Gathering helm status";helm ls -A >> "$DIR/helm_status.log"
+echo "======================Collecting Some General enviornment Information in the $ASTRONOMER_NAMESPACE======================"
+echo "Gathering get all status  in $ASTRONOMER_NAMESPACE Namespace";kubectl get all --all-namespaces > "$DIR/getall_status_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+echo "Gathering All replica status in all namespaces";kubectl get rs --all-namespaces|grep -v '0         0         0' > "$DIR/rs_status_all_namespaces.log"
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:318]: Relying on the amount of whitespace in a grep is not the right way to do this. Consider using `awk '$1 == 0 && $2 == 0 && $3 == 0 {print}'` or something more precise than grep.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+# echo "Gathering kube-system pod status";kubectl get pods -o wide -n kube-system > "$ASTRONOMER_NAMESPACE/Deployment_logs/kube-system.log" 
+# echo "Gathering sevice Status in $ASTRONOMER_NAMESPACE Namespace ";kubectl get svc > "$ASTRONOMER_NAMESPACE/Deployment_logs/svc_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+# echo "Gathering persistent volume Status in $ASTRONOMER_NAMESPACE Namespace ";kubectl get pvc > "$ASTRONOMER_NAMESPACE/Deployment_logs/$PVC_DIR/pvc_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+# echo "Gathering ingress Status in $ASTRONOMER_NAMESPACE Namespace ";kubectl get ingress > "$ASTRONOMER_NAMESPACE/Deployment_logs/ingress_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+# echo "Gathering cronjobs Status in $ASTRONOMER_NAMESPACE Namespace ";kubectl get cronjobs > "$ASTRONOMER_NAMESPACE/Deployment_logs/cronjobs_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+# echo "Gathering jobs Status in $ASTRONOMER_NAMESPACE Namespace ";kubectl get jobs > "$ASTRONOMER_NAMESPACE/Deployment_logs/jobs_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:304]: Why is all this code commented out? We should delete it if it's not being used, or hide it behind an env var or CLI flag if it is sometimes used.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+echo "Gathering logs of prometheus in $ASTRONOMER_NAMESPACE Namespace ";kubectl logs sts/astronomer-prometheus  -c prometheus   > "$ASTRONOMER_NAMESPACE/Deployment_logs/prometheus_container_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+echo "Gathering logs of nats in $ASTRONOMER_NAMESPACE Namespace ";kubectl logs sts/astronomer-nats -c nats  > "$ASTRONOMER_NAMESPACE/Deployment_logs/nats_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+echo "Gathering logs of alertmanager in $ASTRONOMER_NAMESPACE Namespace ";kubectl logs sts/astronomer-alertmanager -c alertmanager  > "$ASTRONOMER_NAMESPACE/Deployment_logs/alertmanager_$ASTRONOMER_NAMESPACE.log" -n $ASTRONOMER_NAMESPACE
+echo "Gathering helm history in $ASTRONOMER_NAMESPACE Namespace";helm history $ASTRONOMER_RELEASE -n $ASTRONOMER_NAMESPACE > "$ASTRONOMER_NAMESPACE/Deployment_logs/helm_history_$ASTRONOMER_RELEASE.log"
+echo "Gathering helm values from $ASTRONOMER_NAMESPACE Namespace";helm get values $ASTRONOMER_RELEASE -n $ASTRONOMER_NAMESPACE -o yaml > "$ASTRONOMER_NAMESPACE/Deployment_logs/helm_values_$ASTRONOMER_RELEASE.yaml"
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:294]: These lines should be replaced with calls to a function that does the thing that is being repeated.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+
+
+
+#####====================================================================================================================================================#####
+# ####Gathering Describe output of all the Nodes
+# echo "======================Gathering Describe output of all the nodes======================"
+# for NODE in $(kubectl get nodes --no-headers |awk '{ print $1}') ; do
+#     echo "Collecting Describe output of Node $NODE ";kubectl describe nodes $NODE > "$ASTRONOMER_NAMESPACE/Deployment_logs/DESCRIBE_$NODE.log"
+# done
+#####==================================================================================================================================================#####
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:270]: This comment block should be removed or put behind an ENV var or command line arg.

Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:270]: These comment blocks should be removed or put behind an ENV var or command line arg.

File: support-utils/cre_astro_software_gather_logs.sh [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+  echo "======================Gathering Ingresses in $NS======================"
+  # Get all ingresses in the namespace
+  kubectl get ingress -n "$NS" > "$NS/${NS}_ingresses.txt"
+
+done
+
+
+ #========================================================== #========================================================== #========================================================== #==========================================================
+
+ #========================================================== #========================================================== #========================================================== #==========================================================
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:190]:
```suggestion
```

File: support-utils/cre_astro_software_gather_logs.sh [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+
+
+# Create a node folder at the same level as the namespace folders
+NODE_DIR="NODES_INFO"
+mkdir -p "$NODE_DIR"
+
+# Get and save the describe output of each node
+for node in $(kubectl get nodes --no-headers | awk '{print $1}'); do
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:198]:
```suggestion
kubectl get nodes --no-headers -o name | while read -r node ; do
```

File: support-utils/cre_astro_software_gather_logs.sh [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+  # Gather logs of all pods
+  for POD in $(kubectl get pods --no-headers -n "$NS" | awk '{print $1}'); do
+    echo "Starting to collect log of the pod $POD in $NS namespace"
+    for container_name in $(kubectl get pods "$POD" -o jsonpath='{.spec.containers[*].name}' -n "$NS" | awk '{NF-=0; OFS="\n"; $1=$1}1' | sort); do
+      echo "Collecting log of the container $container_name in pod $POD in the $NS namespace now"
+      kubectl logs "$POD" -n "$NS" -c "$container_name" > "$NS/AllPodlogs/$POD-pod_$container_name-container.log"
+    done
+  done
+
+  echo "======================Gathering Events in $NS======================"
```
Comment (danielhoherd) [inline support-utils/cre_astro_software_gather_logs.sh:165]: We should probably replace all these `echo "======= whatever ======"` with a function that automatically does the formatting.

File: support-utils/cre_astro_software_gather_logs.sh (hunk 5 above)
Comment (nandlalyadav57) [reply support-utils/cre_astro_software_gather_logs.sh:363]: The intention was to test the responsiveness of the endpoints by sending multiple requests

**Comment by rishkarajgi** [comment]
Code needs cleaning up. Please remove unused/commented code
Also please update the PR description with necessary details of the information sources we are fetching from

File: support-utils/cre_astro_software_gather_logs.sh (hunk 1 above)
Comment (nandlalyadav57) [reply support-utils/cre_astro_software_gather_logs.sh]: used example.com to fix this

**Comment by danielhoherd** [comment]
After discussion about this, we came to the conclusion that this repo is not the right place for this script. I think all the same code review comments apply, and I'd like to see this code cleaned up wherever we end up using it. If you need help with that @nandlalyadav57 ping me on slack.

Here's the slack thread where we are discussing it <https://astronomer.slack.com/archives/C07G6J17AP6/p1723758506894359>

**Pull request closed without merging**